README.rst

```diff
@@ -74,11 +68,140 @@ To install in a virtual environment in your current project:
+Usage Examples
+==============
+
+See `examples/ble_file_transfer_simpletest.py` for a client example. A stub server implementation is in `examples/ble_file_transfer_simpletest.py`.
+
+Protocol
+=========
+
+The file transfer protocol is meant to be simple and easy to implement. It uses free space counts as a way to rate limit file content data transfer. All multi-byte numbers are encoded with the least significant byte first ("<" in CPython's struct module).
+
+GATT Service
+--------------
+
+The UUID of the service is `0xfebb`, Adafruit's 16-bit service UUID.
+
+The base UUID used in characteristics is `ADAFxxxx-4669-6C65-5472-616E73666572`. The 16-bit numbers below are substituted into the `xxxx` portion.
+
+The service has two characteristics:
+
+* version (`0x0100`) - Simple unsigned 32-bit integer version number. Always 1.
+* raw transfer (`0x0200`) - Bidirectional link with a custom protocol. The client does WRITE_NO_RESPONSE to the characteristic and then server replies via NOTIFY. (This is similar to the Nordic UART Service but on a single characteristic rather than two.) The commands over the transfer characteristic are idepotent and stateless. A disconnect during a command will reset the state.
+
+Commands
+---------
+
+Commands always start with a fixed header. The first entry is always the command number itself encoded in a single byte. The number of subsequent entries in the header will vary by command. The entire header must be sent as a unit so set the characteristic with the full header packet.
+
+Paths use `/` as a separator and full paths must start with `/`. Directory paths
+must end with `/` when provided as a full path.
+
+All numbers are unsigned.
+
+Status bytes are `0x81` for OK and `0x82` for error. Other values for error may be used for specific commands.
+
+`0x01` - Read a file
+++++++++++++++++++++
+
+Given a full path, returns the full contents of the file.
+
+The header is three fixed entries and a variable length path:
+
+* Command: Single byte. Always `0x01`.
+* Chunk size: 32-bit number encoding the amount of data that the client can handle in the first reply.
+* Path length: 16-bit number encoding the encoded length of the path string.
+* Path: UTF-8 encoded string that is *not* null terminated. (We send the length instead.)
+
+The server will respond with:
+* Command: Single byte. Always `0x01`.
+* Status: Single byte.
+* Total length: 32-bit number encoding the total file length.
+* Chunk length: 32-bit number encoding the length of the read data up to the chunk size provided in the header.
+* Chunk-length contents of the file starting from the current position.
+
+If the chunk length is smaller than the total length, then the client will request more data by sending:
+* Command: Single byte. Always `0x01`.
+* Status: Single byte. Always OK for now.
+* Chunk size: 32-bit number encoding the number of bytes to read. May be different than the original size. Does not need to be limited by the total size.
```

Comment:
This seems like it assumes errorless transfer and no dropped packets or acknowledgments, etc. I am not sure how well this will work out in the presence of marginal connections or noise. I don't understand how this is idempotent without specifying the position of the chunk. It seems like the position is implied. based on the sizes of the previous chunks ("current position"). Would you also want a checksum, etc.?

Reply:
Ya, the protocol relies on the underlying BLE stack for reliability. This may be a fool's errand but I thought I'd try it first.

By idempotent I mean on a per-command basis (not per-packet.) Maybe it should be per packet... Primarily I wanted to compare against my previous design where two characteristics shared state. One was the file path and that'd influence the state of the second.

Reply:
It's just that if it gets out of sync things will go quite badly. I think passing a file offset for the chunk might make monitoring the stream and debugging problems easier. It will provide something to assert against (not literally `assert()`) when transferring.

I guess a checksum on an already checksummed thing is unnecessary. (The low level has a checksum already, right?)

Reply:
READ and WRITE could be modified not to assume that the full contents are being fetched, but just some chunk, specified as offset and length. That would give you easy partial reads and easy overwrites or appends.

Reply:
SO says there is a CRC: https://stackoverflow.com/questions/44526541/is-a-cyclic-redundancy-check-required-in-a-ble-service

I'll play with the offsets tomorrow. I think it is a good idea. Thanks!